p2p/host/autonat/autonat.go

```diff
@@ -165,117 +170,126 @@ func (as *AmbientAutoNAT) background() {
+	// but fails when the host is private and used with AutoRelay.
+	addrChangeTicker := time.NewTicker(1 * time.Minute)
+	defer addrChangeTicker.Stop()
+
 	timer := time.NewTimer(delay)
 	defer timer.Stop()
 	timerRunning := true
-	retryProbe := false
+	forceProbe := false
 	for {
 		select {
-		// new inbound connection.
 		case conn := <-as.inboundConn:
 			localAddrs := as.host.Addrs()
 			if manet.IsPublicAddr(conn.RemoteMultiaddr()) &&
 				!ipInList(conn.RemoteMultiaddr(), localAddrs) {
 				as.lastInbound = time.Now()
 			}
-
+		case <-addrChangeTicker.C:
+			// schedule a new probe if addresses have changed
 		case e := <-subChan:
 			switch e := e.(type) {
-			case event.EvtLocalAddressesUpdated:
-				// On local address update, reduce confidence from maximum so that we schedule
-				// the next probe sooner
-				if as.confidence == maxConfidence {
-					as.confidence--
-				}
 			case event.EvtPeerIdentificationCompleted:
-				if s, err := as.host.Peerstore().SupportsProtocols(e.Peer, AutoNATProto); err == nil && len(s) > 0 {
-					currentStatus := *as.status.Load()
-					if currentStatus == network.ReachabilityUnknown {
-						as.tryProbe(e.Peer)
-					}
+				if proto, err := as.host.Peerstore().SupportsProtocols(e.Peer, AutoNATProto); err == nil && len(proto) > 0 {
+					forceProbe = true
 				}
+			case event.EvtLocalAddressesUpdated:
+				// schedule a new probe if addresses have changed
 			default:
 				log.Errorf("unknown event type: %T", e)
 			}
-
-		// probe finished.
 		case err, ok := <-as.dialResponses:
 			if !ok {
 				return
 			}
+			as.pendingProbes--
 			if IsDialRefused(err) {
-				retryProbe = true
+				forceProbe = true
 			} else {
 				as.handleDialResponse(err)
 			}
 		case <-timer.C:
+			timerRunning = false
+			forceProbe = false
+			// Update the last probe time. We use it to ensure
+			// that we don't spam the peerstore.
+			as.lastProbe = time.Now()
 			peer := as.getPeerToProbe()
 			as.tryProbe(peer)
-			timerRunning = false
-			retryProbe = false
 		case <-as.ctx.Done():
 			return
 		}
+		// On address update, reduce confidence from maximum so that we schedule
+		// the next probe sooner
+		hasNewAddr := as.checkAddrs()
+		if hasNewAddr && as.confidence == maxConfidence {
+			as.confidence--
+		}
 
-		// Drain the timer channel if it hasn't fired in preparation for Resetting it.
 		if timerRunning && !timer.Stop() {
 			<-timer.C
 		}
-		timer.Reset(as.scheduleProbe(retryProbe))
+		timer.Reset(as.scheduleProbe(forceProbe))
 		timerRunning = true
 	}
 }
 
-func (as *AmbientAutoNAT) cleanupRecentProbes() {
-	fixedNow := time.Now()
-	for k, v := range as.recentProbes {
-		if fixedNow.Sub(v) > as.throttlePeerPeriod {
-			delete(as.recentProbes, k)
+func (as *AmbientAutoNAT) checkAddrs() (hasNewAddr bool) {
+	currentAddrs := as.addressFunc()
+	for _, a := range currentAddrs {
+		if !manet.IsPublicAddr(a) {
+			continue
+		}
+		if _, ok := as.ourAddrs[string(a.Bytes())]; !ok {
```

Comment:
We only need to check if `a` is in the `ourAddrs` set. We don't need the IsPublicAddr function call. Also we can use slices.IndexFunc for a bit of semantic legibility.